core/src/main/java/org/elasticsearch/action/admin/cluster/health/TransportClusterHealthAction.java

```diff
@@ -105,7 +107,7 @@ public void onNoLongerMaster(String source) {
 
                 @Override
                 public void onFailure(String source, Exception e) {
-                    logger.error("unexpected failure during [{}]", e, source);
+                    logger.error((Supplier<?>) () -> new ParameterizedMessage("unexpected failure during [{}]", source), e);
```

Comment:
@rjernst The proposal from @pickypg is for a method

```java
public static Supplier<ParameterizedMessage> message(String message, Object... params) {
  return () -> new ParameterizedMessage(message, params);
}
```

This will cause an allocation of an `Object[]` and all primitive arguments will be boxed to `Object`s. Adding overrides like

```java
public static Supplier<ParameterizedMessage> message(String message, Object p1, Object p2) {
  return () -> new ParameterizedMessage(message, params);
}
```

removes the `Object[]` allocation, but still any primitive arguments will be boxed to `Object`s.

These allocations/boxing will occur regardless of whether or not the log level is fine enough to emit the corresponding log message.

With the `(Supplier<?>) () -> new ParameterizedMessage("{}", o)` pattern (as in this PR) we avoid this boxing unless the log message is going to be emitted.